sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+        return att_names
+
+    def self_att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.self_att_names())
+        return att_names
+
+    def reset(self):
+        for layer in self.layers:
+            layer.reset()
+
+    def num_states(self, step):
+        return sum(layer.num_states(step) for layer in self.layers)
+
+    def state_variables(self, step: int):
+        return [var for layer in self.layers for var in layer.state_variables(step)]
+
+    def state_shapes(self,
+                     batch_size: int,
+                     step: int,
+                     source_encoded_max_length: int,
+                     source_encoded_num_hidden: int):
+        return [state_shape for layer in self.layers for state_shape in layer.state_shapes(batch_size,
+                                                                                           step,
+                                                                                           source_encoded_max_length,
+                                                                                           source_encoded_num_hidden)]
+
+    def init_states(self,
+                    batch_size: int,
+                    source_encoded: List[mx.sym.Symbol],
+                    source_encoded_lengths: mx.sym.Symbol,
+                    source_encoded_max_length: int):
+        return [init_state for layer in self.layers for init_state in layer.init_states(batch_size,
+                                                                                        source_encoded,
+                                                                                        source_encoded_lengths,
+                                                                                        source_encoded_max_length)]
+
+    def get_max_seq_len(self):
+        return min((layer.get_max_seq_len() for layer in self.layers if layer.get_max_seq_len() is not None),
+                   default=None)
+
+
+class EncoderLayerChain(EncoderLayer):
+
+    def __init__(self, layers: List[EncoderLayer]):
+        self.layers = layers
+
+    def encode_sequence(self, source_encoded: mx.sym.Symbol, source_encoded_lengths: mx.sym.Symbol,
+                        source_encoded_max_length: int, att_dict) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
+        for layer in self.layers:
+            source_encoded, source_encoded_lengths, source_encoded_max_length = layer.encode_sequence(
+                source_encoded, source_encoded_lengths, source_encoded_max_length, att_dict)
+        return source_encoded, source_encoded_lengths, source_encoded_max_length
+
+    def att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.att_names())
+        return att_names
+
+    def get_num_hidden(self) -> int:
+        return self.layers[-1].get_num_hidden()
+
+
+class DecoderLayerChain(NestedDecoderLayer):
```

Comment:
maybe move NestedDecoderLayer directly above this class

Reply:
sounds good